### PR TITLE
Implement Fill Inset Rect

### PR DESCRIPTION
**Pull request by duncanspumpkin**
~~I think I'll do the flags for this before ready for review.~~
Heavily based on the OpenRCT2 equivalent but i noticed a few differences when comparing them.

### CODE REVIEW

File: src/OpenLoco/src/Drawing/DrawingContext.h [hunk 1]
```diff
@@ -4,10 +4,23 @@
...
 {
+    enum class DrawRectInsetFlags : uint8_t
+    {
+        fillTransparent = (1u << 2), // ? unused
+        borderNone = (1u << 3),      // ? unused
+        fillNone = (1u << 4),
+        borderInset = (1u << 5),
+        fillDarker = (1u << 6),
+        colourLight = (1u << 7),
+        none = 0u
```
Comment (ZehMatt) [inline src/OpenLoco/src/Drawing/DrawingContext.h:20]: Should be capital U

Comment (ZehMatt) [reply src/OpenLoco/src/Drawing/DrawingContext.h:20]: Also you don't have to put them in parenthesis, don't think we do that.

File: src/OpenLoco/src/Drawing/DrawingContext.h [hunk 2]
```diff
@@ -4,10 +4,23 @@
 #include "Graphics/PaletteMap.h"
 #include "Types.hpp"
 #include "Ui/Rect.h"
+#include <OpenLoco/Core/EnumFlags.hpp>
 #include <cstdint>
 
 namespace OpenLoco::Drawing
 {
+    enum class DrawRectInsetFlags : uint8_t
```
Comment (ZehMatt) [inline src/OpenLoco/src/Drawing/DrawingContext.h:12]: Should we actually call it DrawRectInsetFlags? Considering its in the Drawing namespace we can probably omit the Draw here at least, it makes quite the long lines.

**Comment by duncanspumpkin** [comment]
I've also tacked on at the end a refactor of the many users of drawRect so its much clearer the parameters.

**Review by ZehMatt — APPROVED** [review]

File: src/OpenLoco/src/Drawing/SoftwareDrawingContext.cpp [hunk 3]
```diff
@@ -1497,43 +1497,147 @@ namespace OpenLoco::Drawing
...
-            regs.ax = left;
-            regs.bx = right;
-            regs.cx = top;
-            regs.dx = bottom;
-            regs.ebp = colour;
-            regs.edi = X86Pointer(&rt);
-            regs.si = flags;
-            call(0x004C58C7, regs);
+            const auto rect = Ui::Rect::fromLTRB(left, top, right, bottom);
+            const auto baseColour = static_cast<OpenLoco::Colour>(colour);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Drawing/SoftwareDrawingContext.cpp:1531]: Is the `OpenLoco::` qualifier necessary?

File: src/OpenLoco/src/Drawing/SoftwareDrawingContext.h [hunk 4]
```diff
@@ -104,10 +104,10 @@ namespace OpenLoco::Drawing
...
 
-        void fillRect(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, uint32_t colour) override;
-        void drawRect(Gfx::RenderTarget& rt, int16_t x, int16_t y, uint16_t dx, uint16_t dy, uint32_t colour) override;
-        void fillRectInset(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, uint32_t colour, uint8_t flags) override;
-        void drawRectInset(Gfx::RenderTarget& rt, int16_t x, int16_t y, uint16_t dx, uint16_t dy, uint32_t colour, uint8_t flags) override;
+        void fillRect(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, uint8_t colour, RectFlags flags) override;
+        void drawRect(Gfx::RenderTarget& rt, int16_t x, int16_t y, uint16_t dx, uint16_t dy, uint8_t colour, RectFlags flags) override;
+        void fillRectInset(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, AdvancedColour colour, RectInsetFlags flags) override;
+        void drawRectInset(Gfx::RenderTarget& rt, int16_t x, int16_t y, uint16_t dx, uint16_t dy, AdvancedColour colour, RectInsetFlags flags) override;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Drawing/SoftwareDrawingContext.h:110]: Perhaps we should make `RectFlags::none` the default for these flags parameters?

**Review by AaronVanGeffen — APPROVED** [review]
Left some minor comments for improvement, but looks like a nice cleanup otherwise. Great to get these flags named and strictly typed, too.

File: src/OpenLoco/src/Drawing/SoftwareDrawingContext.h (hunk 4 above)
Comment (ZehMatt) [reply src/OpenLoco/src/Drawing/SoftwareDrawingContext.h:110]: Defaults always have the problem that you might miss the last parameter and it just compiles, I would rather have interface parameters always explicit.

File: src/OpenLoco/src/Drawing/SoftwareDrawingContext.cpp (hunk 3 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/Drawing/SoftwareDrawingContext.cpp:1531]: No i guess not

File: src/OpenLoco/src/Drawing/SoftwareDrawingContext.cpp [hunk 5]
```diff
@@ -1497,43 +1497,147 @@ namespace OpenLoco::Drawing
...
+            // const auto left = rect.left();
+            // const auto top = rect.top();
+            // const auto bottom = rect.bottom();
+            // const auto right = rect.right();
+            // const auto leftTop = Ui::Point32{ rect.left(), rect.top() };
+            // const auto leftBottom = Ui::Point32{ rect.left(), rect.bottom() };
+            // const auto rightTop = Ui::Point32{ rect.right(), rect.top() };
+            // const auto rightBottom = Ui::Point32{ rect.right(), rect.bottom() };
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/OpenLoco/src/Drawing/SoftwareDrawingContext.cpp:1539]: yeah I can't decide if we should convert the parameters to take a Ui::Rect in this PR or another. I think do it in another.

**Review by AaronVanGeffen — APPROVED** [review]

**Pull request merged**